File: JpocToolsCoreCore/CTRL/AbstractCtrl.vb
Imports System.Diagnostics
Public MustInherit Class AbstractCtrl
    Implements IDisposable
    Implements ICtrl

#Region "イベント"
    Public Event OnStep(ByVal sender As Object, ByVal e As OnStepEventArgs)
#End Region

#Region "変数"
    Protected mDto As AbstractDto
    Protected mLogic As AbstractLogic
    Private mDBManager As ElsDataBase
#End Region

#Region "プロパティ"
    Protected ReadOnly Property DBManager() As ElsDataBase
        Get
            Return mDBManager
        End Get
    End Property
#End Region

#Region "コンストラクタ"
    Private Sub New()
        Me.Init()
    End Sub
    Protected Sub New(pDto As AbstractDto)
        Me.New()
        Try
            mDto = pDto
            mDBManager = New ElsDataBase(mDto.ConString)
            Dim objName As String = Me.GetType.ToString
            Dim formId As String = objName.Substring(objName.Length - 6, 6)
            Dim logicType As Type = Type.GetType("Jpoc.Tools.Core.Logic" & formId)
            If logicType IsNot Nothing Then
                Dim obj As Object = Activator.CreateInstance(logicType, mDBManager, mDto)
                If obj IsNot Nothing Then mLogic = CType(obj, AbstractLogic)
            End If
        Catch ex As ElsException
            ex.AddStackFrame(New StackFrame(True))
            Throw
        Catch ex As Exception
            Throw New ElsException(ex)
        End Try
    End Sub
#End Region

#Region "初期化"
    Public MustOverride Sub Init() Implements ICtrl.Init
#End Region

#Region "接続先変更"
    Public Sub ResetConnection()
        Try
            mDBManager.ResetConStr(mDto.ConString)
        Catch ex As ElsException
            ex.AddStackFrame(New StackFrame(True))
            Throw
        Catch ex As Exception
            Throw New ElsException(ex)
        End Try
    End Sub
#End Region

#Region "IDisposable Support"
    Private disposedValue As Boolean ' 重複する呼び出しを検出するには

    ' IDisposable
    Protected Overridable Sub Dispose(disposing As Boolean)
        If Not Me.disposedValue Then
            If disposing Then
                'マネージ状態を破棄します (マネージ オブジェクト)。
                If Me.DBManager IsNot Nothing Then
                    If Me.DBManager.State = ConnectionState.Connecting Then
                        If Me.DBManager.HasTransaction Then Me.DBManager.RollbackTransaction()
                        Me.DBManager.DisConnect()
                    End If
                    Me.DBManager.Dispose()
                End If
            End If

            'アンマネージ リソース (アンマネージ オブジェクト) を解放し、下の Finalize() をオーバーライドします。

            '大きなフィールドを null に設定します。
            Me.mDBManager = Nothing
        End If
        Me.disposedValue = True
    End Sub

    '上の Dispose(ByVal disposing As Boolean) にアンマネージ リソースを解放するコードがある場合にのみ、Finalize() をオーバーライドします。
    'Protected Overrides Sub Finalize()
    '    ' このコードを変更しないでください。クリーンアップ コードを上の Dispose(ByVal disposing As Boolean) に記述します。
    '    Dispose(False)
    '    MyBase.Finalize()
    'End Sub

    ' このコードは、破棄可能なパターンを正しく実装できるように Visual Basic によって追加されました。
    Public Sub Dispose() Implements IDisposable.Dispose, ICtrl.Dispose
        ' このコードを変更しないでください。クリーンアップ コードを上の Dispose(ByVal disposing As Boolean) に記述します。
        Dispose(True)
        GC.SuppressFinalize(Me)
    End Sub
#End Region

#Region "OnStepイベントRaise"
    Protected Sub PassStep(ByVal pStepEventType As PublicEnum.eStepEventType, _
                           ByVal pTarget As String, _
                           ByVal pMessage As String, _
                           Optional ByVal pCompleteStep As Integer = 0, _
                           Optional ByVal pTotalStep As Integer = 0, _
                           Optional ByVal pCompleteSubStep As Integer = 0, _
                           Optional ByVal pTotalSubStep As Integer = 0, _
                           Optional ByVal pCompleteTransactionStep As Integer = 0, _
                           Optional ByVal pTotalTransactionStep As Integer = 0)
        Dim ea As New OnStepEventArgs
        ea.StepEventType = pStepEventType
        ea.Target = pTarget
        ea.Message = pMessage
        ea.CompleteMainStep = pCompleteStep
        ea.TotalMainStep = pTotalStep
        ea.CompleteSubStep = pCompleteSubStep
        ea.TotalSubStep = pTotalSubStep
        ea.CompleteTransactionStep = pCompleteTransactionStep
        ea.TotalTransactionStep = pTotalTransactionStep
        RaiseEvent OnStep(Me, ea)
    End Sub
#End Region

#Region "StepCount加算"
    Protected Function AddStep(ByRef p As Integer) As Integer
        p += 1
        Return p
    End Function
#End Region

End Class

File: JPOC_VM_DEMO/App_Code/CTRL/AbstractCtrl.cs
using Jpoc.Common;
using Jpoc.Dac;
using Jpoc.Tools.Core;
using System.Data;
using System.Diagnostics;

using Utilities = JPOC_VM_DEMO.Common.Utilities;
using GlobalVariables = JPOC_VM_DEMO.Common.GlobalVariables;
using AbstractDto = JPOC_VM_DEMO.App_Code.CTRL.DTO.AbstractDto;
using AbstractLogic = JPOC_VM_DEMO.App_Code.CTRL.Logic.AbstractLogic;
using PublicEnum = JPOC_VM_DEMO.Common.PublicEnum;
using JPOC_VM_DEMO.App_Code.CTRL.Logic;

namespace JPOC_VM_DEMO.App_Code.CTRL
{
    public abstract class AbstractCtrl : IDisposable, ICtrl
    {
        #region Fields
        protected readonly AbstractDto _dto;
        protected AbstractLogic _logic;
        private ElsDataBase _dbManager;
        private bool _disposedValue;
        #endregion

        #region Properties
        protected ElsDataBase DbManager => _dbManager ?? throw new InvalidOperationException("Database manager not initialized");
        protected AbstractDto Dto => _dto;
        protected AbstractLogic Logic => _logic;

        protected string ObjTypeName
        {
            get
            {
                var objName = GetType().ToString();
                var ns = Utilities.GetNameSpace(objName);
                return objName.Replace(ns, string.Empty)
                             .Replace(".Ctrl", string.Empty);
            }
        }
        #endregion

        #region Constructors
        private AbstractCtrl()
        {
            Init();
        }

        protected AbstractCtrl(AbstractDto dto) : this()
        {
            try
            {
                _dto = dto;
                _dbManager = new ElsDataBase(GlobalVariables.ConnectionString);

                // Case sensitive - maintain original behavior
                //var logicName = $"jpoc.Logic{ObjTypeName}";
                var logicName = $"JPOC_VM_DEMO.App_Code.CTRL.Logic.Logic{ObjTypeName}";
                var logicType = Type.GetType(logicName);

                if (logicType != null)
                {
                    var obj = Activator.CreateInstance(logicType, _dbManager, _dto);
                    if (obj != null)
                    {
                        _logic = (AbstractLogic)obj;
                    }
                }
            }
            catch (ElsException ex)
            {
                ex.AddStackFrame(new StackFrame(true));
                throw;
            }
            catch (Exception ex)
            {
                throw new ElsException(ex);
            }
        }
        #endregion

        #region Initialization
        public abstract void Init();
        #endregion

        #region IDisposable Implementation
        protected virtual void Dispose(bool disposing)
        {
            if (!_disposedValue)
            {
                if (disposing)
                {
                    if (_dbManager != null)
                    {
                        if (_dbManager.State == ConnectionState.Connecting)
                        {
                            if (_dbManager.HasTransaction)
                            {
                                _dbManager.RollbackTransaction();
                            }
                            _dbManager.DisConnect();
                        }
                        _dbManager.Dispose();
                    }
                }

                _dbManager = null;
                _disposedValue = true;
            }
        }

        public void Dispose()
        {
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }
        #endregion

        #region Session Management
        public string GetSessionID(int userID, PublicEnum.eUserType userType)
        {
            try
            {
                if (userType == PublicEnum.eUserType.INT)
                {
                    return string.Empty;
                }

                DbManager.Connect();

                //TODO: FIX 

                //return userType == PublicEnum.UserType.INS
                //    ? Logic?.GetSessionIdFromInsUser(userID) ?? string.Empty
                //    : Logic?.GetSessionIdFromUser(userID) ?? string.Empty;
                return string.Empty;
            }
            catch (ElsException ex)
            {
                ex.AddStackFrame(new StackFrame(true));
                throw;
            }
            catch (Exception ex)
            {
                throw new ElsException(ex);
            }
            finally
            {
                DbManager.DisConnect();
            }
        }
        #endregion
    }
}

